common/ledger/util/fileutil/fileutil.go

```diff
@@ -0,0 +1,151 @@
+// the final file will not be visible or it will have the full contents.
+// The tmpFile should not be existing. The finalFile, if exists, will be overwritten (default rename behavior)
+func CreateAndSyncFileAtomically(dir, tmpFile, finalFile string, content []byte, perm os.FileMode) error {
+	tempFilePath := filepath.Join(dir, tmpFile)
+	finalFilePath := filepath.Join(dir, finalFile)
+	if err := CreateAndSyncFile(tempFilePath, content, perm); err != nil {
+		return err
+	}
+	if err := os.Rename(tempFilePath, finalFilePath); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CreateAndSyncFile creates a file, writes the content and syncs the file
+func CreateAndSyncFile(filePath string, content []byte, perm os.FileMode) error {
+	file, err := os.OpenFile(filePath, os.O_RDWR|os.O_CREATE|os.O_EXCL, perm)
+	if err != nil {
+		return errors.Wrapf(err, "error while creating file:%s", filePath)
+	}
+	_, err = file.Write(content)
+	if err != nil {
+		file.Close()
+		return errors.Wrapf(err, "error while writing to file:%s", filePath)
+	}
+	if err = file.Sync(); err != nil {
+		file.Close()
+		return errors.Wrapf(err, "error while synching the file:%s", filePath)
+	}
+	if err := file.Close(); err != nil {
+		return errors.Wrapf(err, "error while closing the file:%s", filePath)
+	}
+	return nil
+}
+
+// SyncDir fsyncs the given dir
+func SyncDir(dirPath string) error {
+	dir, err := os.Open(dirPath)
+	if err != nil {
+		return errors.Wrapf(err, "error while opening dir:%s", dirPath)
+	}
+	if err := dir.Sync(); err != nil {
+		dir.Close()
+		return errors.Wrapf(err, "error while synching dir:%s", dirPath)
+	}
+	if err := dir.Close(); err != nil {
+		return errors.Wrapf(err, "error while closing dir:%s", dirPath)
+	}
+	return err
+}
+
+// SyncParentDir fsyncs the parent dir of the given path
+func SyncParentDir(path string) error {
+	return SyncDir(filepath.Dir(path))
+}
+
+// CreateDirIfMissing makes sure that the dir exists and returns whether the dir is empty
+func CreateDirIfMissing(dirPath string) (bool, error) {
+	if err := os.MkdirAll(dirPath, 0755); err != nil {
+		return false, errors.Wrapf(err, "error while creating dir: %s", dirPath)
+	}
+	if err := SyncParentDir(dirPath); err != nil {
+		return false, err
+	}
+	return DirEmpty(dirPath)
+}
+
+// DirEmpty returns true if the dir at dirPath is empty
+func DirEmpty(dirPath string) (bool, error) {
+	f, err := os.Open(dirPath)
+	if err != nil {
+		logger.Debugf("Error opening dir [%s]: %+v", dirPath, err)
+		return false, errors.Wrapf(err, "error opening dir [%s]", dirPath)
+	}
+	defer f.Close()
+
+	_, err = f.Readdir(1)
+	if err == io.EOF {
+		return true, nil
+	}
+	err = errors.Wrapf(err, "error checking if dir [%s] is empty", dirPath)
+	return false, err
+}
+
+// FileExists checks whether the given file exists.
+// If the file exists, this method also returns the size of the file.
+func FileExists(path string) (bool, int64, error) {
```

Comment:
Really minor but you might want to factor this and `DirExists` together to avoid the 90+% shared code

Reply:
Sure, will have a re-look

Reply:
sorry @ale-linux, I re-looked but there not much scope here.